compiler/rustc_codegen_ssa/src/mir/rvalue.rs

```diff
@@ -143,6 +148,50 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         }
     }
 
+    fn codegen_transmute(
+        &mut self,
+        bx: &mut Bx,
+        src: OperandRef<'tcx, Bx::Value>,
+        dst: PlaceRef<'tcx, Bx::Value>,
+    ) {
+        // The MIR validator enforces no unsized transmutes.
+        debug_assert!(src.layout.is_sized());
+        debug_assert!(dst.layout.is_sized());
+
+        if src.layout.size != dst.layout.size
+            || src.layout.abi == abi::Abi::Uninhabited
+            || dst.layout.abi == abi::Abi::Uninhabited
+        {
+            // In all of these cases it's UB to run this transmute, but that's
+            // known statically so might as well trap for it, rather than just
+            // making it unreachable.
+            bx.abort();
+            return;
+        }
+
+        let size_in_bytes = src.layout.size.bytes();
+        if size_in_bytes == 0 {
+            // Nothing to write
+            return;
+        }
+
+        if let OperandValue::Ref(src_llval, meta, src_align) = src.val {
+            debug_assert_eq!(meta, None);
+            // For a place-to-place transmute, call `memcpy` directly so that
+            // both arguments get the best-available alignment information.
+            let bytes = bx.cx().const_usize(size_in_bytes);
+            let flags = MemFlags::empty();
+            bx.memcpy(dst.llval, dst.align, src_llval, src_align, bytes, flags);
+        } else {
+            // Otherwise we have an immediate (or a pair of them), so pointer-cast
+            // the destination to pretend it's has the same layout as source.
+            let llty = bx.backend_type(src.layout);
+            let cast_ptr = bx.pointercast(dst.llval, bx.type_ptr_to(llty));
+            let align = src.layout.align.abi.min(dst.align);
```

Comment:
Why can't we just give this the destination's alignment?

Reply:
Oh, good point!  It was like this before so that the `memcpy` it did internally would be correct, but now that it's only immediates that's no longer relevant.